### PR TITLE
Build system overhaul.

### PR DESCRIPTION
**Pull request by bardiharborow**
This PR switches every build task to a NPM script, *except* saucelabs (this is surprisingly non-trivial, and I'm pushing it to a later PR as this has taken way too long as it is). Closes #20332.

List of tasks that need to included. A ticked checkbox indicates that the task has been tested.

- [x]  CSS
  - [x] Linting (scss-lint)
    - [x] Core
    - [x] Docs
  - [x] Compile (node-sass)
    - [x] Core
    - [x] Docs
  - [x] Prefix (postcss)
    - [x] Core
    - [x] Docs
  - [x] Minify (cleancss)
    - [x] Core
    - [x] Docs
- [x] JavaScript
  - [x] Linting (eslint)
    - [x] Core
    - [x] Docs
  - [x] Concatenate + Transpile (Babel)
    - [x] Bundle
    - [x] Individual plugins
  - [x] Minify (uglifyjs)
    - [x] Core
    - [x] Docs
  - [x] Test (qunit + PhantomJS)
    - [x] Local
    - [x] SauceLabs (punting conversion to npm)
- [x]  Docs
  - [x] Clean (`cp -r dist/* docs/dist/`)
  - [x] Linting (htmlhint + htmllint)
  - [x] Compile (Jekyll)
    - [x] Local
    - [x] GitHub
  - [x] Serve (Jekyll)
  - [x] <strike>Push to GitHub</strike> (punting)
- [x] Maintenance
  - [x] Update dependencies (npm-check-update)
  - [x] Shrinkwrap (npm)
- [x] Release
  - [x] Change version number (custom node script)
  - [x] Package zip file (GNU zip)
  - [x] <strike>Push to npm, etc.</strike> (punting)

### CODE REVIEW

File: js/tests/index.html [hunk 1]
```diff
@@ -2,138 +2,94 @@
...
-      QUnit.done(function (testResults) {
-        var tests = []
+
+      // See https://github.com/axemclion/grunt-saucelabs#test-result-details-with-qunit
+      var log = [];
+
+      QUnit.done(function(testResults) {
+        testResults.tests = [];
+
         for (var i = 0, len = log.length; i < len; i++) {
```
Comment (Johann-S) [inline js/tests/index.html]: Why did you do : ``len = log.length; i < len`` instead of ``i < log.length`` ?

Comment (bardiharborow) [inline js/tests/index.html]: Oh, if you look carefully, I haven't changed that line. Caching the length of an array used to speed the loop up a little, but it appears [Chrome does this optimisation automatically now](https://stackoverflow.com/questions/5752906/is-reading-the-length-property-of-an-array-really-that-expensive-an-operation), so we can actually remove it. I've changes to the test loader off into #22047.

**Comment by mdo** [comment]
How best can I give feedback on this @bardiharborow? Do you have a checklist of remaining todos, or specific things that need testing?

**Comment by bardiharborow** [comment]
@mdo I've added a list of task and their review status. A local `npm test` now works (finally!). I have to work on replacing `grunt-stamp`, `grunt-saucelabs` and `grunt-build-control`, and then we might actually be done!

**Comment by mdo** [comment]
`grunt-build-control` can probably go away. I don't use it right now and just manually copy-pasta our generated docs and stuff :). I'd love something that actual worked, but in the spirit of moving this forward, we can do that later.

**Comment by hanakin** [comment]
Pls consider using stylelint over scss lint as it offers far more control and flexibility in code control with plugin support for additional control

**Comment by bardiharborow** [comment]
@hanakin We'll be shipping that after this PR merges. It's just going to require a lot of config wrangling, and I'm juggling enough changes in this PR as it is.

**Comment by bardiharborow** [comment]
Preparing to ship this. I have every expectation that something will go wrong and I'll be thoroughly [trouted](https://en.wiktionary.org/wiki/trouted) for it, but anyway...

I'll ship this at **2:00am UTC on 20th of April 2017**, when I'll be awake and able to fix any incidents. Please ping me on GitHub and email me if I fail to respond.

File: package.json [hunk 2]
```diff
@@ -17,24 +17,40 @@
...
+    "css-prefix": "postcss --config build/ --replace dist/css/*.css",
+    "css-prefix-docs": "postcss --config build/ --no-map --replace docs/assets/css/docs.min.css docs/examples/**/*.css",
+    "css-minify": "cleancss --level 1 --source-map --output dist/css/bootstrap.min.css dist/css/bootstrap.css && cleancss --level 1 --source-map --output dist/css/bootstrap-grid.min.css dist/css/bootstrap-grid.css && cleancss --level 1 --source-map --output dist/css/bootstrap-reboot.min.css dist/css/bootstrap-reboot.css",
+    "css-minify-docs": "cleancss --level 1 --source-map --output docs/assets/css/docs.min.css docs/assets/css/docs.min.css",
+    "js": "npm-run-all js-compile js-minify",
+    "js-docs": "npm-run-all js-lint-docs js-minify-docs",
+    "js-lint": "eslint js/ && eslint --config js/tests/.eslintrc.json --env node build/",
+    "js-lint-docs": "eslint --config js/tests/.eslintrc.json docs/assets/js/",
+    "js-compile": "npm-run-all --parallel js-compile-*",
+    "js-compile-bundle": "shx cat js/src/util.js 'js/src/!(util|popover).js' js/src/popover.js | shx sed 's/^(import|export).*//' | babel --filename js/src/bootstrap.js | node build/stamp.js > dist/js/bootstrap.js",
```
Comment (Johann-S) [inline package.json]: Hi @bardiharborow, 

I have an error when this task is runned see : 

image

I know my screenshot is in french, but it said : 
`'popover).js'' isn't known at an intern or extern command`

Comment (bardiharborow) [inline package.json]: Okay. I thought that shx was performing the glob expansion, but maybe bash is doing it, and that obviously won't work on Windows. I'll replace this with a static list for now.

**Comment by patrickhlauke** [comment]
not sure if it's something broken at my end, but: i fetched and rebased my fork on my machine, did a new `npm install`...but now when i try to run `grunt` i simply get `Warning: Task "default" not found. Use --force to continue.`

**Comment by Johann-S** [comment]
They are just one task on our [Gruntfile.js](https://github.com/twbs/bootstrap/blob/v4-dev/Gruntfile.js) the other are in [package.json](https://github.com/twbs/bootstrap/blob/v4-dev/package.json)

And for example to run the `css` task you have to do : 
`npm run css`

**Comment by patrickhlauke** [comment]
so forgive me being thick, but...what's the equivalent now of the old `grunt` which built everything, followed by `bundle exec jekyll serve`

**Comment by Johann-S** [comment]
It seems it's `test` task and after you can call `docs-serve` task which will call `bundle exec jekyll serve`

**Comment by patrickhlauke** [comment]
tried `npm run test` on a hunch, and it errors out with "Could not find jekyll-sitemap-1.1.1 in any of the sources". joys of running this on windows...

**Comment by Johann-S** [comment]
Yes it seems they are several bugs on Windows (I'm on Windows too) a PR is open here #22477 we should report any issues here

**Comment by mdo** [comment]
Awesome to see this merged! You're amazing @bardiharborow. Super happy to see things consolidated :). Only hiccup has been some Sauce failures with running the Sauce grunt task. See #22395 and #22496. Unsure what's going on there just yet.

**Comment by silverwind** [comment]
Would it make sense to have a `npm run dist`? I've been using the old `grunt dist` to just build CSS+JS, and found this to be roughly equal in the new build system (thought it does not do parallelism):

````
npm-run-all css-compile css-prefix css-minify js-compile-bundle js-compile-plugins js-minify
````

**Pull request merged**